db/column_family.cc

```diff
@@ -529,14 +529,30 @@ std::unique_ptr<WriteControllerToken> SetupDelay(
     // insert to mem tables, so we need to actively slow down before we get
     // feedback signal from compaction and flushes to avoid the full stop
     // because of hitting the max write buffer number.
-    if (prev_compaction_neeed_bytes > 0 &&
-        prev_compaction_neeed_bytes <= compaction_needed_bytes) {
+    //
+    // If DB just falled into the stop condition, we need to further reduce
+    // the write rate to avoid the stop condition.
+    if (near_stop) {
+      // Penalize the near stop condition with trippling the slowdown.
+      // This is to provide the long term slowdown increase signal.
+      // The penalty is more than the reward of recovering to the normal
+      // condition.
+      write_rate = static_cast<uint64_t>(
+          static_cast<double>(write_rate) /
+          (kSlowdownRatio * kSlowdownRatio * kSlowdownRatio));
+      if (write_rate < kMinWriteRate) {
+        write_rate = kMinWriteRate;
+      }
+    }
+    if (was_stopped ||
```

Comment:
We don't reduce write rate when we fall into stop, because we don't determine delay value if stop condition hits. Here we pay the debt. Also I want to penalize going to stop condition. This is something we want to avoid.